File: eternal_return/src/main/java/org/EternalReturn/ERCharacter/ERCharacter.java
package org.EternalReturn.ERCharacter;

import org.EternalReturn.ERCharacter.GlobalMonobehav.PlayerRayCastingByLeftClicking;
import org.EternalReturn.EREntity.EREntity;
import org.EternalReturn.ERPlayer.ERPlayer;
import org.EternalReturn.System.PluginInstance;
import org.bukkit.entity.Player;

import java.util.HashMap;
import java.util.Map;

public abstract class ERCharacter extends EREntity {

    protected ERPlayer erPlayer;

    public long ActiveCooldownSeconds;
    public long PassiveCooldownSeconds;
    public SkillCooldown cooldown = new SkillCooldown();

    public ERCharacter(ERPlayer erPlayer){
        super(PluginInstance.getEREngine().createOrientedBox(erPlayer.getPlayer().getLocation(),1.0/2,3.0/2,1.0/2));
        this.setEntity(erPlayer.getPlayer());
        this.erPlayer = erPlayer;
        registerMonobehaviour(new PlayerRayCastingByLeftClicking());
    }

    public abstract String getName();

    public ERPlayer getERPlayer(){
        return this.erPlayer;
    }

    public Player getPlayer(){
        return (Player)this.getEntity();
    }

    public SkillCooldown getCooldown() {
        return this.cooldown;
    }
}

File: eternal_return/src/main/kotlin/org/EternalReturn/ERCharacter/ERCharacter.kt
package org.EternalReturn.ERCharacter

import org.EternalReturn.ERCharacter.globalmonobehav.PlayerRayCastingByLeftClicking
import org.EternalReturn.ERCharacter.globalmonobehav.PlayerRayCastingMeleeAttack
import org.EternalReturn.EREntity.EREntity
import org.EternalReturn.ERPlayer.ERPlayer
import org.EternalReturn.System.PluginInstance
import org.bukkit.entity.Player

abstract class ERCharacter(player : Player, erPlayer: ERPlayer) : EREntity(
    PluginInstance.getEREngine().createOrientedBox(erPlayer.player.location, 1.0 / 2, 3.0 / 2, 1.0 / 2),
    player
) {
    var eRPlayer: ERPlayer
        protected set

    @JvmField
    var ActiveCooldownSeconds: Long = 0
    @JvmField
    var PassiveCooldownSeconds: Long = 0
    var cooldown: SkillCooldown = SkillCooldown()

    init {
        this.entity = player
        this.eRPlayer = erPlayer

        registerMonobehaviour(PlayerRayCastingByLeftClicking())
        registerMonobehaviour(PlayerRayCastingMeleeAttack())
    }

    abstract val name: String

    val player: Player
        get() = this.entity as Player
}
